### PR TITLE
Fix Jump forever, Feature collect coins, Feature collect score

### PR DESCRIPTION
**Pull request by serafinaliu**

### CODE REVIEW

**Review by Aidy199258 — APPROVED** [review]

File: src/com/ECS/Component.java [hunk 1]
```diff
@@ -20,14 +20,22 @@
 }
 
 
-class CoinComponent extends Component{
```
Comment (serafinaliu) [inline src/com/ECS/Component.java:23]: Updated Coin Component data type.

File: src/com/ECS/System.java [hunk 2]
```diff
@@ -455,6 +455,23 @@ public void KeyPressed(ArrayList<Entity> entities, KeyEvent e) {
 //Counting points/scores
 //Incomplete
 class ScoreSystem extends System{
+    public static boolean CatchCoin(ArrayList<Entity> entities, float x, float y){
+        for(Entity entity: entities){
+            if(entity.hasComponent(CoinComponent.class)) {
```
Comment (serafinaliu) [inline src/com/ECS/System.java:460]: Look for Coin

File: src/com/ECS/System.java [hunk 3]
```diff
@@ -455,6 +455,23 @@ public void KeyPressed(ArrayList<Entity> entities, KeyEvent e) {
 //Counting points/scores
 //Incomplete
 class ScoreSystem extends System{
+    public static boolean CatchCoin(ArrayList<Entity> entities, float x, float y){
```
Comment (serafinaliu) [inline src/com/ECS/System.java:458]: x and y are the player X, Y here

File: src/com/ECS/System.java [hunk 4]
```diff
@@ -455,6 +455,23 @@ public void KeyPressed(ArrayList<Entity> entities, KeyEvent e) {
...
+        for(Entity entity: entities){
+            if(entity.hasComponent(CoinComponent.class)) {
+                CoinComponent coin = (CoinComponent)entity.getComponent(CoinComponent.class);
+                if (coin.CoinPoint > 0) {
+                    float range = 50;
+                    if ((x - range) < coin.getX() && coin.getX() < (x + range)) {
+                        if ((y - range) < coin.getY() && coin.getY() < (y + range)) {
```
Comment (serafinaliu) [inline src/com/ECS/System.java:465]: Check if Player is in the same location as the coin

File: src/com/ECS/System.java [hunk 5]
```diff
@@ -455,6 +455,23 @@ public void KeyPressed(ArrayList<Entity> entities, KeyEvent e) {
...
+                            coin.CoinPoint = 0;
```
Comment (serafinaliu) [inline src/com/ECS/System.java:466]: remove the point from Coin if Player catch that coin

File: src/com/ECS/System.java [hunk 6]
```diff
@@ -773,23 +789,30 @@ public void Process(ArrayList<Entity> entities, double dt) {
...
+                boolean catchCoin = ScoreSystem.CatchCoin(entities, position.getX(), position.getY());
+                if (catchCoin){
+                    point.GainOnePoint();
+                }
+
                 //java.lang.System.out.println("Velocity: " + velocity.Get_VelocityX());
 
 
-                if(gravity == true) {
+                if(gravity == true && velocity.Get_VelocityY() == 0) {
```
Comment (serafinaliu) [inline src/com/ECS/System.java:808]: only allow player jump, when the last jump ended

File: src/com/ECS/System.java [hunk 7]
```diff
@@ -859,8 +882,25 @@ public static void LoadPicturesToEntity1(Entity entity,Image img){
 
     public void Process(ArrayList<Entity> entities, Graphics2D g) {
         for (Entity entity : entities) {
-            // Need a Position & a Render Component
-            if(entity.hasComponent(PositionComponent.class) && entity.hasComponent(RenderComponent.class)) {
+            if (entity.hasComponent(CoinComponent.class) && entity.hasComponent(RenderComponent.class)) {
```
Comment (serafinaliu) [inline src/com/ECS/System.java:885]: New Rendering logic for display the Coin on Screen

File: src/com/ECS/System.java [hunk 8]
```diff
@@ -859,8 +882,25 @@ public static void LoadPicturesToEntity1(Entity entity,Image img){
...
+                CoinComponent coin = (CoinComponent)entity.getComponent(CoinComponent.class);
+                if (coin.CoinPoint > 0) {
```
Comment (serafinaliu) [inline src/com/ECS/System.java:887]: Only display the coin if there is point remaining in the coin.

**Pull request merged**